blocks/log.py

```diff
@@ -1,274 +1,418 @@
 """The event-based main loop of Blocks."""
+import binascii
+import datetime
+import os
+# import sqlite3
 from abc import ABCMeta, abstractmethod
-from collections import defaultdict
+from collections import Mapping, MutableMapping, OrderedDict
+from numbers import Integral
+from operator import itemgetter
 
-from six import add_metaclass
+import numpy
+import six
+from picklable_itertools import imap
 try:
-    from pandas import DataFrame
-    PANDAS_AVAILABLE = True
+    from bson.objectid import ObjectId
+    from pymongo import MongoClient
+    PYMONGO_AVAILABLE = True
 except ImportError:
-    PANDAS_AVAILABLE = False
+    PYMONGO_AVAILABLE = False
 
 
-@add_metaclass(ABCMeta)
-class AbstractTrainingStatus(object):
-    """The base class for objects that carry the training status.
+class TrainingLog(Mapping):
+    r"""The log of training progress.
 
-    To support various backends (such as database tables) the
-    descendants of this class are expected to override the
-    `__getattr__`, `__setattr__`, `__hasattr__` methods.
+    A training log stores the training timeline, statistics and other
+    auxiliary information. Information is represented as nested dictionary
+    of iteration, key values.
 
-    Attributes
-    ----------
-    iterations_done : int
-        The number of iterations done.
-    epochs_done : int
-        The number of epochs done.
-    _epoch_ends : list
-        The numbers of the epochs last iterations.
+    In addition to the set of records of the training progress, a training
+    log has a status object whose attributes describe the current state of
+    training e.g. whether training has started, whether termination has
+    been requested, etc.
 
-    .. todo::
+    Parameters
+    ----------
+    hash_ : str, optional
```

Comment:
The idea is--although not entirely worked out yet--that in order to load a log stored in the database, you can instantiate it by passing the correct hash.